src/main/java/net/snowflake/ingest/streaming/internal/InternalStageManager.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2024 Snowflake Computing Inc. All rights reserved.
+ */
+
+package net.snowflake.ingest.streaming.internal;
+
+import static net.snowflake.ingest.utils.Constants.BLOB_EXTENSION_TYPE;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.util.Calendar;
+import java.util.TimeZone;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import net.snowflake.client.jdbc.SnowflakeSQLException;
+import net.snowflake.ingest.utils.ErrorCode;
+import net.snowflake.ingest.utils.SFException;
+import net.snowflake.ingest.utils.Utils;
+
+/** Class to manage single Snowflake internal stage */
+class InternalStageManager<T> implements StorageManager<T> {
+  // Target stage for the client
+  private final StreamingIngestStorage<T> targetStage;
+
+  // Increasing counter to generate a unique blob name per client
+  private final AtomicLong counter;
+
+  // Whether the manager in test mode
+  private final boolean isTestMode;
+
+  // Snowflake service client used for configure calls
+  private final SnowflakeServiceClient snowflakeServiceClient;
+
+  // Client prefix generated by the Snowflake server
+  private final String clientPrefix;
+
+  /**
+   * Constructor for InternalStageManager
+   *
+   * @param isTestMode whether the manager in test mode
+   * @param client the owning client
+   * @param snowflakeServiceClient the Snowflake service client to use for configure calls
+   */
+  InternalStageManager(
+      boolean isTestMode,
+      SnowflakeStreamingIngestClientInternal<T> client,
```

Comment:
This client is being passed into stageManager purely for name and role. What if we passed those as explicit fields on this constructor? instead of coupling together these two classes and introducing a cyclic dependency between the classes?
InternalStageManager is a "lower layer" class that ideally should not depend on a "higher layer" class like the StreamingIngestClient.

Reply:
Make sense, exposed `role` and `clientName` in new function parameters.